### PR TITLE
Removes the QTX dialect

### PR DESCRIPTION
**Pull request by boschmitt**
### Description
This PR removes the QTX dialect and related infrastructure. Two translated passes were adapted to work with Quake instead of QTX and the common dialect is almost removed (next PR will finish it).

### CODE REVIEW

File: include/cudaq/Optimizer/CodeGen/Passes.td [hunk 1]
```diff
@@ -12,11 +12,11 @@
 include "mlir/Pass/PassBase.td"
 
 def QuakeToQIR : Pass<"quake-to-qir", "mlir::ModuleOp"> {
-  let summary = "Lower Quake/QTX to QIR.";
+  let summary = "Lower Quake to QIR.";
   let description = [{
-    Converts Quake (or QTX) to QIR (as LLVM-IR).
-    The quake-add-deallocs pass should be run before this pass
-    in order to properly generate deallocations for allocated qubits.
+    Converts Quak to QIR (as LLVM-IR). The quake-add-deallocs pass should be run
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/CodeGen/Passes.td:17]:
```suggestion
    Converts Quake to QIR (as LLVM-IR). The quake-add-deallocs pass should be run
```

File: include/cudaq/Optimizer/Transforms/Passes.td [hunk 2]
```diff
@@ -299,79 +296,4 @@ def UnwindLowering : Pass<"unwind-lowering", "mlir::func::FuncOp"> {
...
-    construction rule(s).  Decomposition techniques are faster than synthesis
-    and produce predictable results, i.e., we know in advance the cost in number
-    of low-level instructions and in the number of qubits.
-
-    NOTE: The implementation still preliminary.
-  }];
-  let constructor = "cudaq::opt::createOpDecompositionPass()";
-}
-
-def SplitArrays: Pass<"qtx-split-arrays", "qtx::CircuitOp"> {
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Transforms/Passes.td:333]: A pass similar to this probably needs to be implemented in the unified dialect, right?

File: include/cudaq/Target/OpenQASM/OpenQASMEmitter.h [hunk 3]
```diff
@@ -14,8 +14,9 @@
 
 namespace cudaq {
 
-/// Translates the given operation to OpenQASM code. The operation or operations
-/// in the region of 'op' need almost all be in QTX dialect.
+/// Translates the given operation to OpenQASM code. The operation, `op`,  or
+/// operations in its region must be in Quake memory reference form. Also,
+/// qubit vectors cannot have unknown size.
```
Comment (schweitzpgi) [inline include/cudaq/Target/OpenQASM/OpenQASMEmitter.h:19]: FYI: I've been eliminating the use a "qubit" in the IR documentation and instead using "quantum reference" and/or "wire" as appropriate. "qubit" seemed a bit too slippery in the IR. (Admittedly, the ambiguity might be exactly what we want here?)

File: lib/Target/IQM/TranslateToIQMJson.cpp [hunk 4]
```diff
@@ -21,16 +21,17 @@ namespace cudaq {
...
+
 static LogicalResult emitEntryPoint(nlohmann::json &json, Emitter &emitter,
-                                    qtx::CircuitOp circuitOp) {
-  if (circuitOp.getBody().getBlocks().size() != 1)
-    circuitOp.emitError("Cannot map qtx Circuit op with more than 1 block to "
-                        "IQM Json. Must be a flat circuit representation.");
+                                    func::FuncOp op) {
+  if (op.getBody().getBlocks().size() != 1)
+    op.emitError("Cannot translate kernels with more than 1 block to IQM Json. "
+                 "Must be a flat representation.");
```
Comment (schweitzpgi) [inline lib/Target/IQM/TranslateToIQMJson.cpp:29]:
```suggestion
                 "Must be a straight-line representation.");
```
I'm thinking this means no special control flow (ops with regions) as well.

File: lib/Target/OpenQASM/TranslateToOpenQASM.cpp [hunk 5]
```diff
@@ -116,72 +111,94 @@ static LogicalResult emitOperation(Emitter &emitter, ModuleOp moduleOp) {
...
-static LogicalResult emitOperation(Emitter &emitter, qtx::CircuitOp circuitOp) {
-  // Check if we can translate
-  if (circuitOp.getClassicalResultTypes().size())
-    return circuitOp.emitError("cannot return classical results");
-  for (auto target : circuitOp.getTargets()) {
-    if (isa<qtx::WireArrayType>(target.getType()))
-      return circuitOp.emitError(
-          "Cannot translate array arguments into OpenQASM 2.0");
+static LogicalResult emitOperation(Emitter &emitter, func::FuncOp op) {
+  // In quake, kernels only return classical types. Thus, we check whether the
```
Comment (schweitzpgi) [inline lib/Target/OpenQASM/TranslateToOpenQASM.cpp:164]: It'll be permitted to transform a device kernel into "pass-by-value" form with wire arguments and wire results. There isn't a pass that does that, but we can certainly write it. In fact, I'd like to see if the !quake.control type can't be leveraged as well with such a transformation to help detect circuit bugs.

File: lib/Target/OpenQASM/TranslateToOpenQASM.cpp [hunk 6]
```diff
@@ -116,72 +111,94 @@ static LogicalResult emitOperation(Emitter &emitter, ModuleOp moduleOp) {
...
+  // numbers of results is zero or not.
+  if (op.getNumResults() > 0)
+    return op.emitError("cannot return classical results");
+
+  // Separate classical and quantum arguments.
+  SmallVector<Value, 4> parameters;
+  SmallVector<Value, 4> targets;
```
Comment (schweitzpgi) [inline lib/Target/OpenQASM/TranslateToOpenQASM.cpp:171]: nit: The `4` length is not needed.

**Review by schweitzpgi — COMMENTED** [review]
LGTM. Thanks for helping clean this up, Bruno. Greatly appreciated.

BTW, I'm planning to steal the gist of the Conversion tests in the memtoreg/regtomem PR that's in the pipeline.

**Review by amccaskey — APPROVED** [review]

**Pull request merged**